src/main/java/software/amazon/qldb/QldbDriverImpl.java

```diff
@@ -171,24 +171,22 @@ public void execute(ExecutorNoReturn executor, RetryPolicy retryPolicy) {
                 retryAttempt++;
                 logger.info("A recoverable error has occurred. Attempting retry #{}.", retryAttempt);
                 logger.debug("Errored Transaction ID: {}. Error cause: ", ee.getTransactionId(), ee.getCause());
-                if (ee.isInvalidSessionException()) {
-                    logger.debug("Replacing expired session...");
-                    replaceDeadSession = true;
+                replaceDeadSession = !ee.isAbortSuccessful();
+                if (replaceDeadSession) {
+                    logger.debug("Replacing invalid session...");
                 } else {
                     logger.debug("Retrying with a different session...");
-                    replaceDeadSession = false;
-                    if (ee.isAborted()) {
-                        this.releaseSession(session);
-                    } else {
-                        this.poolPermits.release();
-                    }
+                    this.releaseSession(session);
                 }
 
                 try {
-                    RetryPolicyContext context = new RetryPolicyContext(ee.getCause(), retryAttempt, ee.getTransactionId());
+                    // This can be safely casted since only SdkExceptions or child Exceptions are deemed retryable.
+                    SdkException se = (SdkException)ee.getCause();
```

Comment:
We should override getCause() in ExecuteException to return an SdkException so this isn't necessary.

Reply:
Actually, I see that's already done so this should be a spurious cast.